src/mono/mono/mini/interp/transform.c

```diff
@@ -2459,13 +2459,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 			*op = MINT_MONO_MEMORY_BARRIER;
 		else if (!strcmp (tm, "Exchange") && csignature->param_count == 2 && csignature->params [0]->type == MONO_TYPE_I8 && csignature->params [1]->type == MONO_TYPE_I8)
 			*op = MINT_MONO_EXCHANGE_I8;
-		else if (!strcmp (tm, "CompareExchange") && csignature->param_count == 3 &&
-			 (csignature->params[1]->type == MONO_TYPE_I4 ||
-			  csignature->params[1]->type == MONO_TYPE_I8)) {
-			if (csignature->params[1]->type == MONO_TYPE_I4)
-				*op = MINT_MONO_CMPXCHG_I4;
-			else
-				*op = MINT_MONO_CMPXCHG_I8;
+		else if (!strcmp (tm, "CompareExchange") && csignature->param_count == 3) {
+			MonoType *t = mini_get_underlying_type (csignature->params[1]);
+			if (t->type == MONO_TYPE_I4 ||
```

Comment:
do you really need to guard the switch with an if?

Reply:
probably not. I wasn't sure if I would make signed/unsigned opcodes or if I would emit a second instruction to fixup the result operand.  in the end it seemed better to have signed/unsigned opcodes.  I'll remove the "if" once I see a green x64 run